// cronJobs/lib/jobManagementDbConnector.js
var _ = require('lodash')
    , Shared = require('mia-js-core/node_modules/shared')
    , ServerHeartbeatModel = require('./cronServerHeartbeatModel.js')
    , CronJobExecutionModel = require('./cronJobExecutionModel.js')
    , Utils = require('mia-js-core/node_modules/utils')
    , Encryption = Utils.Encryption
    , Q = require('q');

function thisModule() {
    var self = this;

    var _statusNoHeartbeat = 'no heartbeat';
    var _statusActive = 'active';

    self.registerNewServer = function (hostName, restartCount) {
        var now = new Date();
        return ServerHeartbeatModel.insert({
            status: _statusActive,
            hostName: hostName,
            startedAt: now,
            isSuspended: false,
            restartCount: restartCount,
            lastStatusUpdateAt: now
        }).then(function (serverInfo) {
            if (serverInfo) {
                return Q(serverInfo[0]);
            }
            else {
                return Q.reject('Cannot insert new server info for crons into DB. Crons will be disabled on this server.');
            }
        });
    };

    self.validateDbServerEntry = function (serverUniqueId) {
        return ServerHeartbeatModel.findOne({
            _id: serverUniqueId,
            status: _statusActive,
            isSuspended: {$ne: true}
        }).then(function (serverInfo) {
            return serverInfo ? Q() : Q.reject();
        });
    };

    self.getDbServerEntry = function (serverUniqueId) {
        return ServerHeartbeatModel.findOne({
            _id: serverUniqueId
        });
    };

    self.doHeartbeat = function (serverId) {
        return ServerHeartbeatModel.findAndModify({
            _id: serverId,
            status: _statusActive
        }, {}, {
            $set: {
                'lastStatusUpdateAt': new Date()
            }
        }, {
            new: true
        }).then(function (serverInfo) {
            if (!serverInfo) {
                return Q.reject('heartbeat not possible');
            }
            else {
                return Q(serverInfo);
            }
        });
    };

    self.getOrCreateDbJobEntry = function (jobTypeName, cronConfig) {
        if (cronConfig.allowedHosts == null || cronConfig.allowedHosts.length == 0) {
            cronConfig.allowedHosts = Shared.config('environment.cronJobs.allowedHosts');
        }

        return _getOrCreateDbJobEntry(jobTypeName, {
            config: cronConfig,
            runningJobsByServer: []
        });
    };

    self.getDbJobEntry = function (jobTypeName) {
        return CronJobExecutionModel.findOne({
            typeName: jobTypeName
        });
    };

    var _getOrCreateDbJobEntry = function (jobTypeName, initializationDoc) {
        return CronJobExecutionModel.findOne({
            typeName: jobTypeName
        }).then(function (jobInfo) {
            if (!jobInfo) {
                var doc = _.clone(initializationDoc);
                doc.typeName = jobTypeName;
                return CronJobExecutionModel.insert(doc).then(function (jobInfo) {
                    if (jobInfo) {
                        return Q(jobInfo[0]);
                    }
                    else {
                        return Q.reject('Error inserting job description to db');
                    }
                });
            }
            else {
                return Q(jobInfo);
            }
        });
    };

    self.addServerEntryToJobDocument = function (jobTypeName, serverId) {
        return CronJobExecutionModel.findAndModify({
            typeName: jobTypeName,
            runningJobsByServer: {
                $not: {
                    $elemMatch: {
                        serverId: serverId
                    }
                }
            }
        }, {}, {
            $push: {
                runningJobsByServer: {
                    serverId: serverId,
                    jobs: []
                }
            }
        }, {
            new: true
        });
    };

    self.startNewJob = function (jobTypeName, serverId, hostName, maxJobsPerServer, maxJobsTotal) {
        var jobId = Encryption.randHash();
        return self.addServerEntryToJobDocument(jobTypeName, serverId).then(function () {
            var query = {
                $and: [{
                    typeName: jobTypeName,
                    'config.isSuspended': false,
                    runningJobsByServer: {
                        $elemMatch: {
                            serverId: serverId
                        }
                    }
                }, {
                    $or: [{
                        'config.allowedHosts': hostName
                    }, {
                        'config.allowedHosts': {$exists: false}
                    }, {
                        'config.allowedHosts': {$size: 0}
                    }]
                }]
            };

            if (maxJobsPerServer && maxJobsPerServer != 0) {
                var arraySizeQueryPerServer = 'jobs.' + (maxJobsPerServer - 1);
                query['$and'][0]['runningJobsByServer']['$elemMatch'][arraySizeQueryPerServer] = {$exists: false};
            }

            if (maxJobsTotal && maxJobsTotal != 0) {
                var arraySizeQueryTotal = 'runningJobsOverall.' + (maxJobsTotal - 1);
                query['$and'][0][arraySizeQueryTotal] = {$exists: false};
            }

            var now = new Date();
            return CronJobExecutionModel.findAndModify(query, {}, {
                $set: {
                    'stats.lastStarted.serverId': serverId,
                    'stats.lastStarted.hostName': hostName,
                    'stats.lastStarted.time': now
                },
                $inc: {
                    'stats.timesStarted': 1
                },
                $push: {
                    "runningJobsByServer.$.jobs": {
                        id: jobId
                    },
                    runningJobsOverall: {
                        id: jobId,
                        serverId: serverId,
                        startedAt: now
                    }
                }
            }, {
                new: true
            }).then(function (jobDoc) {
                if (!jobDoc) {
                    return Q.reject('Could not start job');
                }
                else {
                    return Q({jobId: jobId, jobDoc: jobDoc});
                }
            });
        });
    };

    self.stopJob = function (jobId, timeLapsed) {
        return CronJobExecutionModel.update({
            'runningJobsByServer.jobs.id': jobId
        }, {
            $inc: {
                'stats.timesFinished': 1,
                'stats.totalExecutionTime': timeLapsed
            },
            $pull: {
                runningJobsOverall: {
                    id: jobId
                },
                'runningJobsByServer.$.jobs': {
                    id: jobId
                }
            }
        }, {
            multi: true,
            validate: false
        });
    };

    self.cleanJobsOfDeadServers = function (heartbeatInterval) {
        return _identifyDeadServers(heartbeatInterval).then(_killAllDeadJobs);
    };

    var _identifyDeadServers = function (heartbeatInterval) {
        var dateTo = new Date(Date.now() - heartbeatInterval);
        return ServerHeartbeatModel.update({
            status: _statusActive,
            lastStatusUpdateAt: {$lt: dateTo}
        }, {
            $set: {
                status: _statusNoHeartbeat,
                lastStatusUpdateAt: new Date()
            }
        }, {
            multi: true,
            validate: false
        });
    };

    var _killAllDeadJobs = function () {
        return ServerHeartbeatModel.find({
            status: _statusNoHeartbeat
        }).then(function (cursor) {
            return Q.ninvoke(cursor, 'toArray').then(function (servers) {
                return Q(servers.map(function (value) {
                    return _removeAllJobsOfServer(value._id).then(function () {
                        return _removeServer(value._id).then(function () {
                            return Q(value._id);
                        });
                    });
                }));
            });
        }).then(function (promises) {
            return Q.allSettled(promises);
        });
    };

    var _removeServer = function (serverId) {
        return ServerHeartbeatModel.remove({
            _id: serverId
        });
    };

    var _removeAllJobsOfServer = function (serverId) {
        return CronJobExecutionModel.update({}, {
            $pull: {
                runningJobsOverall: {
                    serverId: serverId
                },
                runningJobsByServer: {
                    serverId: serverId
                }
            }
        }, {
            multi: true,
            validate: false,
            w: 1
        });
    };

    return self;
};

module.exports = new thisModule();
